accounting/src/delta/delta_data_collection/tests/delta_delta_delta_tests.rs

```diff
@@ -0,0 +1,107 @@
+#[case(new_delta(Some('a'), None),      new_delta(None,      Some('b')), new_delta(Some('b'), Some('c')), new_delta(Some('a'), Some('c')))]
+#[case(new_delta(Some('a'), Some('b')), new_delta(Some('b'), Some('c')), new_delta(Some('c'), None),      new_delta(Some('a'), None))]
+#[case(new_delta(Some('a'), Some('b')), new_delta(Some('b'), None),      new_delta(None, Some('c')),      new_delta(Some('a'), Some('c')))]
+fn delta_delta_delta_associativity(
+    #[case] delta1: DataDelta<char>,
+    #[case] delta2: DataDelta<char>,
+    #[case] delta3: DataDelta<char>,
+    #[case] expected_delta: DataDelta<char>,
+) {
+    let expected_collection = DeltaDataCollection::from_iter([(1, expected_delta)]);
+
+    {
+        // (Delta + Delta) + Delta = Delta
+        // every delta goes into separate collection
+        let mut collection1 = DeltaDataCollection::from_iter([(1, delta1.clone())]);
+        let collection2 = DeltaDataCollection::from_iter([(1, delta2.clone())]);
+        let collection3 = DeltaDataCollection::from_iter([(1, delta3.clone())]);
+        let _ = collection1.merge_delta_data(collection2).unwrap();
+        let _ = collection1.merge_delta_data(collection3).unwrap();
+
+        assert_eq!(collection1, expected_collection);
+    }
+
+    {
+        // Delta + (Delta + Delta) = Delta
+        // every delta goes into separate collection
+        let mut collection1 = DeltaDataCollection::from_iter([(1, delta1.clone())]);
+        let mut collection2 = DeltaDataCollection::from_iter([(1, delta2.clone())]);
+        let collection3 = DeltaDataCollection::from_iter([(1, delta3.clone())]);
+        let _ = collection2.merge_delta_data(collection3).unwrap();
+        let _ = collection1.merge_delta_data(collection2).unwrap();
+
+        assert_eq!(collection1, expected_collection);
+    }
+
+    {
+        // (Delta + Delta) + Delta = Delta
+        // every delta is applied to the same collection
+        let mut collection = DeltaDataCollection::new();
+        let _ = collection.merge_delta_data_element(1, delta1).unwrap();
+        let _ = collection.merge_delta_data_element(1, delta2).unwrap();
+        let _ = collection.merge_delta_data_element(1, delta3).unwrap();
+
+        assert_eq!(collection, expected_collection);
+    }
+}
+
+proptest! {
+    // This test verifies that combination of 3 random deltas is associative.
+    // Invalid combinations can be generated, but the associativity property
+    // doesn't apply for such cases as different Error can be produced depending on the
+    // order of operations. So for the sake of this test errors are treated as None
+    // and it is only expected that both sequences must produce either the same valid result
+    // or both fail with some error.
+    #[test]
+    fn random_delta_associativity(
+        delta1 in any::<DataDelta<char>>(),
+        delta2 in any::<DataDelta<char>>(),
+        delta3 in any::<DataDelta<char>>(),
```

Comment:
If I recall correctly, if you use the `any` generator, there is a shortcut where you can say just `delta1: DataDelta<char>` etc. It follows more closely the usual function syntax. It's up to you whether you deem it nicer or not.

Reply:
indeed, thanks